src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -721,14 +721,59 @@ export class TerminalInstance implements ITerminalInstance {
 		}
 		if (this._processManager) {
 			this._processManager.dispose(immediate);
+		} else {
+			// In cases where there is no associated process (for example executing an exetnsion callback task)
+			// consumers still expect on onExit event to be fired. An example of this is terminating the extnesion callback
+			// task.
+			this._onExit.fire(this._id);
 		}
+
 		if (!this._isDisposed) {
 			this._isDisposed = true;
 			this._onDisposed.fire(this);
 		}
 		this._disposables = lifecycle.dispose(this._disposables);
 	}
 
+	public finishedWithRenderer(): void {
+		//NOTE: This looks a lot like _onProcessExit, and it should.
+		// The use of this API is for cases where there is no backing process
+		// behind a terminal instance (such as when executiong an extension callback task).
+		// There is no associated string, error text, etc, as the consumer of the renderer
+		// can simply outout the text to the renderer themselves.
+		// All this code does is handle the "wait on exit" condition.
+		if (!this.shellLaunchConfig.isRendererOnly) {
+			return;
+		}
+
+		// Prevent dispose functions being triggered multiple times
+		if (this._isExiting) {
+			return;
+		}
+
+		this._isExiting = true;
+
+		// Only trigger wait on exit when the exit was *not* triggered by the
+		// user (via the `workbench.action.terminal.kill` command).
+		if (this._shellLaunchConfig.waitOnExit) {
+			if (typeof this._shellLaunchConfig.waitOnExit === 'string') {
+				let message = this._shellLaunchConfig.waitOnExit;
+				// Bold the message and add an extra new line to make it stand out from the rest of the output
+				message = `\n\x1b[1m${message}\x1b[0m`;
+				this._xterm.writeln(message);
+			}
+			// Disable all input if the terminal is exiting and listen for next keypress
+			this._xterm.setOption('disableStdin', true);
+			if (this._xterm.textarea) {
+				this._attachPressAnyKeyToCloseListener();
+			}
+
+			this._onExit.fire(this._id);
```

Comment:
Looks like this if is identical to some code in _onProcessExit and could be shared.

Reply:
Yep. Agreed. Unified the code.